### DISCUSSION TITLE
Allow filtering in stored JSON objects

### DISCUSSION THREAD

**Discussion by rijkvanzanten** [directus/directus#7277] [Feature Requests]
Some databases, like Postgres and more recent versions of MySQL have (limited) capabilities to search _in_ JSON objects, instead of treating them as a flat value. Postgres f.e. has a dedicated syntax for it, where MySQL has a "EXTRACT_JSON" function that can be used to extract values from the JSON object.

Seeing this would require a lot of custom magic per DB driver, I think this could be implemented in a similar fashion as the date extraction functions (`year(datetime_created)` etc):

```
json(my_json_field, path.to.the.value)
```

The second argument would be a dot-notation path of where to find the value in the JSON, similar to `lodash`es `get` function

**Comment by casualmatt** [comment]
Because now we can filter records by a json field with "is null" or "isn't null" and is a bit annoying and frustrating.

**Comment by rasco** [comment]
this feature would be great.

json functions are already in Knex, so it probably wouldn't require custom magic per DB driver: http://knexjs.org/#Builder-whereJsonObject

**Comment by tspvivek** [comment]
@rijkvanzanten Is there any workaround or hack to filter in JSON objects now?

**Reply by rijkvanzanten** [reply]
In a custom hook/endpoint, you can access the knex instance used to communicate with the database, at which point you could run some custom SQL to add filtering capabilities based on your specific database vendor.

**Reply by tspvivek** [reply]
Thanks. Will try this.

**Reply by rasco** [reply]
@tspvivek take a look at this. maybe you don't need SQL: http://knexjs.org/#Builder-whereJsonObject

**Reply by piotr-cz** [reply]
I believe new link is: http://knexjs.org/guide/schema-builder.html#json

**Comment by hermesalvesbr** [comment]
Any chance of this coming back to the API?

**Reply by rijkvanzanten** [reply]
Coming back? What do you mean? I don't think Directus has ever supported filtering in JSON values 🤔

**Reply by hermesalvesbr** [reply]
/items/noticia/?filter[tag][_contains]="agricultores"

```
{
    "errors": [
        {
            "message": "select \"noticia\".\"conteudo\", \"noticia\".\"date_created\", \"noticia\".\"date_updated\", \"noticia\".\"destaque\", \"noticia\".\"id\", \"noticia\".\"prefeitura\", \"noticia\".\"slug\", \"noticia\".\"sort\", \"noticia\".\"status\", \"noticia\".\"tag\", \"noticia\".\"titulo\", \"noticia\".\"user_created\", \"noticia\".\"url_de_origem\" from \"noticia\" where \"noticia\".\"tag\" like $1 order by \"noticia\".\"sort\" asc limit $2 - operator does not exist: json ~~ unknown",
            "extensions": {
                "code": "INTERNAL_SERVER_ERROR"
            }
        }
    ]
}
```

**Reply by hermesalvesbr** [reply]
![image](https://user-images.githubusercontent.com/379850/183494686-9b247b09-991f-4eb5-8ad4-d07587b1d86c.png)

**Reply by madc** [reply]
See here: https://github.com/directus/directus/issues/14894#issuecomment-1208200622

**Reply by hermesalvesbr** [reply]
> See here: [#14894 (comment)](https://github.com/directus/directus/issues/14894#issuecomment-1208200622)

manually convert all content or create a endpoint :( I saw that.

**Reply by rijkvanzanten** [reply]
@hermesalvesbr `contains` is the operator that will find a substring in a piece of text. It has never worked on JSON fields in Postgres (which is what I'm assuming you're using based on the error message), and was never designed for use with JSON fields in the first place 👍🏻

**Reply by hermesalvesbr** [reply]
I understood. But, from api, his another way to filter a json? Just search all collection?

**Comment by madc** [comment]
This issue just got a little bump via #14894.

**Comment by AaronBeaudoin** [comment]
Just want to resurface this discussion/issue because it's been **since [May 2022](https://github.com/directus/directus/discussions/13418)** that I discovered it also affect's **built-in** Directus tables as well as being a user limitation.

 In short, the **Tags** field in the Directus file library is essentially useless. This is because the purpose of tags is generally understood to be to enable searching/filtering items based on whether a given tag is present, but since the tags field in the Directus file library is (presumably) a JSON object, we **can't** actually use it for searching/filtering. When you try filter on it you get the options "Is null" and "Isn't null", both of which make no sense in the context of tags.

<img width="527" alt="image" src="https://user-images.githubusercontent.com/26592486/169354595-79f9bb29-8935-4786-9e0b-511346b841a6.png">

I'd honestly argue that the **Tags** field should just be _removed entirely_ until this issue is resolved, because right now what exactly would anyone use it for other than searching/filtering as described above? I think this issue is more of a problem than it initially appeared. Alternatively, the **Tags** field could be converted to an actual database relation, but I think the more helpful direction is to focus on resolving this issue and bring the benefits of JSON filtering to everyone.

**Reply by connorwinston** [reply]
JSON querying is in the works right now, check out the progress at https://github.com/directus/directus/pull/15889

**Reply by FINDarkside** [reply]
Is there some reason file tags don't use CSV format instead of JSON? Because it seems like the CSV one is searchable while JSON is not.

**Reply by rijkvanzanten** [reply]
CSV is searchable because it's just a string of text. The JSON filtering aims to add other more useful filtering operations like 'one of', or 'all of these'

**Comment by slysiou** [comment]
Great to see the support on Rest API. But, when do you plan to add it on UI ?
Thanks

**Reply by rijkvanzanten** [reply]
It'll come to the UI as well, but I don't have a timeline for it 👍🏻

**Reply by snimavat** [reply]
Is there any example how posts could be filtered by tags ? through directus sdk ? @rijkvanzanten 

Thnx

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by tspvivek** [reply]
Its already part of https://github.com/directus/directus/pull/15889, right?

**Reply by rijkvanzanten** [reply]
Yeah! Bear with me here while we cleanup all the open feature requests 😄 There's a handful current in under-review that are already (partially) done or otherwise in-flight 🙂

**Comment by w0kyj** [comment]
Data Abstraction Project dependent:
- use relational nested notation to reference the structure
- works for any data source
- how is understood --> add RFC to ensure details defined

**Comment by phazonoverload** [comment]
On January 11 we're inviting you to come and have a chat with us about this feature request at our Request Review in our Discord server. This will help us have some dedicated time to talk about what implementation would have the most sense. Please feel free to join us as https://directus.chat (the event is already listed and you can set a reminder)

**Comment by w0kyj** [comment]
Feature Request Session Notes
Filtering Qs:
- deep/nested filtering of a JSON path within the object
- filter/search arrays
- general text search of the "JSON" blob
- 

Tim: I divided it up in 3 different functionalities in that old #15889 
- fields extracting data from a json field
- filter using specific json fields and filter against that
    - filter for the fields(key), then filter on the values of those fields
- deep actually filter inside the json blob

Is the idea to integrate Directus's existing filters/aggregate/Dynamic Variables function or would there be new JSON specific functions?
- existing filtering for sure
- may include specific filter operators as needed same as other data types today

psql allows for dateranges (a JSON array) and I built a custom daterange interface/display for this but can't filter with it. Is that something that would be supported by this pr or would that be something else?(unsure of other DBs)
- Yes this would be supported by the new feature

Could you consider using JSON schema to type and validate fields.
- interesting idea, need to think through
    - schema for a specific field
    - schema for a collection
    - schema for a item
    - Need a way to define schema
- data validation/enforcement based on the defined schema
    - great for page builder use case -- use json field for page content and other table fields for routes, dates, status, etc
    - repeater interface
    - relational repeater interface
        - no guarantees that data is a key or value
        - Mongo does this with special "type" object id that is globally unique
        - performance will be a key consideration here (some vendors may support, while others won't)
        - This will be a separate feature discussion (it is very tricky)
        - 
- There is a Portable Text spec https://github.com/portabletext/portabletext
-

**Comment by jofmi** [comment]
Hi! Is there a timeline for this feature to also come to the UI? :)

**Reply by phazonoverload** [reply]
We don't offer timelines for any work that we do, sorry.

**Comment by wfpaisa** [comment]
greetings, I am also interested in this issue, I hope it is resolved soon, as I see that the issue has been going on for several years.

**Comment by snimavat** [comment]
Is there any update on it, i see some PRs were done/merged...

**Comment by Nebula-Developer** [comment]
Any updates?

**Comment by neobarthold** [comment]
Any ETA here?

**Reply by rijkvanzanten** [reply]
No timeline yet. It's actively being worked on though 🙂 Hoping we have something ready enough to be announced after Christmas

**Reply by ludekvodicka** [reply]
hello, is there any updated ETA please? I just figured out that we are not able to filter via tags. Should I rework it somehow or is it close to release, please?

**Reply by Jdelbarcogarza** [reply]
Yes, is there any updates on array filtering in directus?

**Comment by BorisKamp** [comment]
Watching this as well, I personally do not understand why there would be a `tags` field type if the APIs cannot filter on the values.

**Comment by BluetileEntertainment** [comment]
I migrated a ton of data from Drupal to Directus that heavily used tags , and then I realized it doesn't support filtering 
I am watching as well, You guys are awesome hope to see this get resolved

**Comment by deepslam** [comment]
Hello guys!

It's just terrible that directus doesn't support this filter! I created a scheme, converted a lot of content and cannot filter them! Are you serious? )))

**Reply by xgenvn** [reply]
you can always model or create a computed for searching. Such as many requests here are asking for tags, you can create a trigger to transform the data and allow it to be filtered. Similar to other object properties or nested fields. In the end, you still need to know SQL and leverage other functions.

**Comment by Yakito** [comment]
I’m trying to understand if I’m missing something. If I create a field using the “tag” type, I can’t filter by a tag. Is that correct? If so, I’d like to bump this thread because it would be great to be able to filter by elements inside those fields.

<img width="186" alt="image" src="https://github.com/user-attachments/assets/05ce2050-c309-4c0f-b955-992e3e42b76f" />

**Reply by emrahnazif** [reply]
> I’m trying to understand if I’m missing something. If I create a field using the “tag” type, I can’t filter by a tag. Is that correct? If so, I’d like to bump this thread because it would be great to be able to filter by elements inside those fields.
> 
> <img alt="image" width="186" src="https://private-user-images.githubusercontent.com/992405/422504155-05ce2050-c309-4c0f-b955-992e3e42b76f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUyNDAxMDMsIm5iZiI6MTc0NTIzOTgwMywicGF0aCI6Ii85OTI0MDUvNDIyNTA0MTU1LTA1Y2UyMDUwLWMzMDktNGMwZi1iOTU1LTk5MmUzZTQyYjc2Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyMVQxMjUwMDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNDYxMzg5Njg2Y2JjYTNlM2M3YzhlNTEzNjAyZDJmMjQ1MTYzYWM1NTM1YTZkYjg1ZjNjZjZjOTVkOWVkN2E4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.4ak63_d2P6oT1wkAQNR90J5Pmtujgn7MwTBmJ09Px-w">

For tags specifically, there are some workarounds, such as using CSV instead of JSON. But this doesn't diminish the importance of filtering JSON objects. Many people are also waiting impatiently for a native solution.

**Comment by dpitch** [comment]
guys, will it be possible to filter tags ?

**Comment by chrney** [comment]
Old thread, I am very aware of this. Just a short check if this still is worked on, or if this has been dismissed overall. I'd be grateful for a short indication on where this is, or is not, heading. Thanks a lot!

**Reply by xluckydegen** [reply]
It would be great to know this. I believe it would be very useful to have this.

**Comment by adiazc9** [comment]
Isn't there any solution yet to allow filtering in stored JSON objects? This feature would be widely apreciated!

**Comment by xgenvn** [comment]
I think either we allow the JSON filtering or allow customizing searching and filtering for collection config, maybe open some spaces to develop extension. Because for some specific needs of searching we want to leverage Directus as middle layer and push data to other search engine.

**Reply by xgenvn** [reply]
If we allow linking the filter and search with workflows, that would be also helpful.

**Comment by flipswitchingmonkey** [comment]
would adding something like `jq` as a middleware into the filter be a viable option?

**Reply by rijkvanzanten** [reply]
Directus' filtering works by generating SQL queries that include the filter, rather than post-processing all the data in the database. Using something like `jq` to post-process filter the works in theory, but in practice it means you have to load all the data from the table into memory

**Comment by JamesW1** [comment]
Hi All,

Filtering in JSON was added recently in the 12.0.0-rc.1 release, sharing reference documentation below:

- [Quickstart](https://directus.com/docs/guides/connect/json/quickstart)
- [Advanced Querying](https://directus.com/docs/guides/connect/json/advanced-querying)

While support has been adding in REST/GraphQL/SDK, we haven't yet added support in the Studio - but do plan to make this available.

If anyone has any feedback, please share it via our public roadmap tool 🙏
https://roadmap.directus.io/request

**Reply by JamesW1** [reply]
Update on this - JSON filtering inside the Studio has landed on the v12.2.0 - for any JSON type field, you will now see the function option in the filter dropdown:

<img width="655" height="335" alt="image" src="https://github.com/user-attachments/assets/a1f0f6d3-c832-4e85-93bf-b55e7fa6668c" />
